packages/tunnel-client/src/lib/tunnel.ts

```diff
@@ -131,25 +144,42 @@ export class Tunnel extends (EventEmitter as new () => TypedEmitter<TunnelEvents
           cb(null, getInfo(body as CreateTunnelResponse));
         })
         .catch((err) => {
-          this.logger.debug(`tunnel server offline: ${err.message}, retry 1s`);
+          this.logger.error(`tunnel server offline: ${err.message}, retry 1s`);
           return setTimeout(getUrl, 1000);
         });
     };
 
     getUrl();
   }
 
-  _establish(info: TunnelInfo) {
+  async _establish(info: TunnelInfo) {
     // increase max event listeners so that localtunnel consumers don't get
     // warning messages as soon as they setup even one listener. See #71
     this.setMaxListeners(
       info.maxConn + (EventEmitter.defaultMaxListeners || 10)
     );
 
-    this.tunnelCluster = new TunnelCluster({ ...info, logger: this.logger });
+    let tunnelCluster: TunnelConnectionCluster | TunnelMultiplexingCluster;
+
+    if (info.multiplexingRemotePort) {
+      this.logger.debug("using multiplexing agent");
+      tunnelCluster = await this._establishMultiplexingCluster({
+        ...info,
+        multiplexingRemotePort: info.multiplexingRemotePort,
+      });
+    } else if (info.remotePort) {
+      this.logger.debug("using connection agent");
+      tunnelCluster = new TunnelConnectionCluster({
+        ...info,
+        remotePort: info.remotePort,
+        logger: this.logger,
+      });
```

Comment:
Why do we need to have extra logic before creating the `TunnelMultiplexingCluster` but we can create the `TunnelConnectionCluster` directly?

Reply:
There is a slight difference in separation of concern between the two. For `TunnelConnectionCluster` because we open a new TCP connection each time we call `open()` we do auth within the "tunnel cluster" but because `TunnelMultiplexingCluster` works over a single connection I've made it so it depends on an open, authenticated TPC connection over which it creates logical duplexes for communication . I guess an alternative might be to open a connection when TunnelMultiplexingCluster is created. I've opted for the former but do shout if you think an alternative here is cleaner!